tests/common/helpers/parallel.py

```diff
@@ -61,69 +73,138 @@ def parallel_run(target, args, kwargs, nodes, timeout=None):
         dict: An instance of multiprocessing.Manager().dict(). It is a proxy to the shared dict that is used by all the
             spawned processes.
     """
+    nodes = [node for node in nodes_list]
+    # Callback API for wait_procs
+    def on_terminate(worker):
+        logger.info("process {} terminated with exit code {}".format(
+            worker.name, worker.returncode)
+        )
+
+    def force_terminate(workers):
+        # Some processes cannot be terminated. Try to kill them and raise flag.
+        running_processes = [worker for worker in workers if worker.is_alive()]
+        if len(running_processes) > 0:
+            logger.info(
+                'Found processes still running: {}. Try to kill them.'.format( #lgtm [py/clear-text-logging-sensitive-data]
+                    str(running_processes)
+                )
+            )
+            for p in running_processes:
+                results[p.name] = [{'failed': True}]
+                try:
+                    os.kill(p.pid, signal.SIGKILL)
+                except OSError as err:
+                    logger.debug("Unable to kill {}:{}, error:{}".format(
+                        p.pid, p.name, err
+                    ))
+                    pass
+
+            pt_assert(
+                False,
+                """Processes running target "{}" could not be terminated.
+                Tried killing them. But please check""".format(target.__name__)
+            )
+
+
     workers = []
     results = Manager().dict()
     start_time = datetime.datetime.now()
-    for node in nodes:
-        kwargs['node'] = node
-        kwargs['results'] = results
-        process_name = "{}--{}".format(target.__name__, node)
-        worker = SonicProcess(name=process_name, target=target, args=args, kwargs=kwargs)
-        worker.start()
-        logger.debug('Started process {} running target "{}"'.format(worker.pid, process_name))
-        workers.append(worker)
+    tasks_done = 0
+    total_tasks = len(nodes)
+    tasks_running = 0
+    total_timeout = timeout * (int(len(nodes)/concurrent_tasks) + 1) if timeout else None
```

Comment:
I think this algorithm is still incorrect. If len(nodes) quals concurrent_tasks, `total_timeout` will be `timeout * 2`.
ceil(4/3) is 1. But ceil(4/3.0) will be 2.

Reply:
I see, float changes the behavior, updated.